### ISSUE TITLE
Azure IoT Edge V2 - Deployment is not fetched by device

### ISSUE THREAD

**Issue by oliverchristen** [microsoft/azure-pipelines-tasks#13412]
**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureIoTEdgeV2

## Environment
- Server - Azure Pipelines or TFS on-premises?
    Azure pipelines

- Agent - Hosted or Private: 
    Hosted agent

## Issue Description

I'm not sure, whether this is a bug in the pipeline task or one in the IoT Hub. But I don't know where to report bugs in IoT Hub, so I'm reporting it here.
When the action `Deploy to IoT Edge devices` is executed, everything seems to be ok. Any existing deployment is overridden on the IoT Hub. But sometimes randomly the deployment is not assigned to the matching devices. In the device the _IoT Edge Runtime Response_ is **417 -- The device's deployment configuration is not set** and in the tab _Deployments_ there is no deployment assigned any more.

I figured out two workarounds which both involve manual actions:

- Delete the deployment on the IoT Hub before releasing a new one
- Edit the deployment (do dummy changes) and save it

In both cases the deployment get fetched by device.

**Comment by oliverchristen** [comment]
The issue still exists. Unfortunately I did not hear anything of @blackchoey.

**Comment by ds-ms** [comment]
Looping in the new owners: @marianan   @tameraw   @pprovost could you guys take a look?

**Comment by marianan** [comment]
@ds-ms thanks for reporting this issue. Could you please try to use the az cli command directly in your pipeline and observe if the issue reproduces - this will help root cause the issue. The az cli command that the task uses is at https://github.com/microsoft/azure-pipelines-tasks/blob/693e719b4dbf83dc3637a040273035de9bb5221e/Tasks/AzureIoTEdgeV2/deployimage.ts#L83.

**Comment by oliverchristen** [comment]
@marianan Thanks for your suggestion. I'll try it out when I have the time to do so.

**Comment by oliverchristen** [comment]
@marianan Today I had the chance to change our release pipeline to use Azure CLI directly instead of IoT Edge task. Until now the problem did not occur again. But it also happened very rarely before. So that does not mean much yet.

But a huge benefit of this change is the increase in performance. Before it took about 90 seconds and now about 30 seconds to deploy the file. I also created an issue for that: https://github.com/microsoft/azure-pipelines-tasks/issues/13977

**Comment by oliverchristen** [comment]
@marianan I was a little bit too fast with my feedback. The exact same problem occurred again even if I now deploy the configuration with Azure CLI. So that means it has to be a problem on IoT Hub, does it?

**Comment by marianan** [comment]
@oliverchristen my next guess would be that this may be related to IoT Hub. I am looking to find the right repo/contacts for next steps in this investigations.

**Comment by and-rewsmith** [comment]
@oliverchristen How often does this repro for you? I am attempting to investigate this from the edge side. I think you are probably correct in that it is a hub issue, so there is someone from the hub team assisting me on that front. 

I feel the most efficient way to resolve this is to get a repro with a way to observe:
- any possible client errors
- hub behavior
- edge reaction to missing "no deployment assigned"

Please take a look at this issue and see if it is similar to your case. They were able to fix it through changing the deployment id.
https://github.com/Azure/iotedge/issues/2450

**Comment by and-rewsmith** [comment]
@oliverchristen Can you also please tell me what version of iotedge service, EdgeHub, and EdgeAgent you are using?

**Comment by oliverchristen** [comment]
@and-rewsmith I'm in contact with the Azure support too.

What I wrote to them:
What I found out:
-	The 417 runtime response is disappearing always after a few minutes
-	The 417 runtime response is reported while new module images are pulled

So for me it seems to be different to the mentioned github issue (https://github.com/Azure/iotedge/issues/2450). There the error appears after a while and do not disappear again. It’s just the opposite as in my case, isn’t it?

It happens about every 5. time I think. I think it depends how many and how large module images have to be updated.

In the edgeHub log it seems that it looses its device configuration but get it again after a while.

So to me everything seems to work as expected except during those few minutes. During this period of time the runtime status is wrong (417) and the deployment assignment is not done.
